### PR TITLE
Support priorityFee fetching from RPC and some better logging

### PR DESCRIPTION
**Pull request by nambrot**
### Description

Like title says, this PR allows the SmartProvider to fetch feeData from the RPC, if supported. Otherwise it falls back to the existing 1.5 gwei hardcoded value that ethers.js v5 has. On most L2s, that is a gross overestimate, resulting in so much wasted ETH.

### Drive-by changes

Some additional logging

### Backward compatibility

Yes

### Testing
Manual

### CODE REVIEW

File: typescript/cli/src/deploy/warp.ts [hunk 1]
```diff
@@ -142,9 +142,12 @@ async function executeDeploy(params: DeployParams) {
     ismFactoryDeployer,
   );
 
+  logGreen('✅ ISM deployments complete');
+  log(`Deploying Warp contracts`);
+
   const deployedContracts = await deployer.deploy(modifiedConfig);
 
-  logGreen('✅ Hyp token deployments complete');
+  logGreen('✅ Warp contract deployments complete');
```
Comment (jmrossy) [inline typescript/cli/src/deploy/warp.ts:150]:
```suggestion
  logGreen('✅ Warp Route contract deployments complete');
```

File: typescript/sdk/src/providers/SmartProvider/HyperlaneJsonRpcProvider.ts [hunk 2]
```diff
@@ -151,6 +158,10 @@ export class HyperlaneJsonRpcProvider
     return combinedResults;
   }
 
+  get maxPriorityFeePerGas(): Promise<BigNumber> {
```
Comment (jmrossy) [inline typescript/sdk/src/providers/SmartProvider/HyperlaneJsonRpcProvider.ts:161]: Avoid `get/set` class member keywords.

File: typescript/sdk/src/providers/SmartProvider/HyperlaneJsonRpcProvider.ts [hunk 3]
```diff
@@ -32,6 +32,13 @@ export class HyperlaneJsonRpcProvider
     super(rpcConfig.connection ?? rpcConfig.http, network);
   }
 
+  prepareRequest(method: string, params: any): [string, any[]] {
+    if (method === ProviderMethod.MaxPriorityFeePerGas) {
+      return ['eth_maxPriorityFeePerGas', []];
+    }
```
Comment (jmrossy) [inline typescript/sdk/src/providers/SmartProvider/HyperlaneJsonRpcProvider.ts:38]: Why special case this?

File: typescript/sdk/src/providers/SmartProvider/SmartProvider.ts [hunk 4]
```diff
@@ -1,4 +1,5 @@
-import { providers } from 'ethers';
+import { BigNumber, providers } from 'ethers';
+import { resolveProperties } from 'ethers/lib/utils.js';
```
Comment (jmrossy) [inline typescript/sdk/src/providers/SmartProvider/SmartProvider.ts:2]: incorrect import. This will screw up tree shaking and can cause other problems. I would guess that importing `utils` and then calling `utils.resolveProperties` is what you're looking for

File: typescript/sdk/src/providers/SmartProvider/SmartProvider.ts [hunk 5]
```diff
@@ -97,6 +98,41 @@ export class HyperlaneSmartProvider
     this.supportedMethods = [...supportedMethods.values()];
   }
 
+  async getPriorityFee() {
+    try {
+      return BigNumber.from(await this.perform('maxPriorityFeePerGas', {}));
+    } catch (error) {
+      return BigNumber.from('1500000000');
```
Comment (jmrossy) [inline typescript/sdk/src/providers/SmartProvider/SmartProvider.ts:105]: The catch will prevent the retry logic. But if this is the logic you want anyway, add logging and move the number to a const

File: typescript/sdk/src/providers/SmartProvider/SmartProvider.ts [hunk 6]
```diff
@@ -97,6 +98,41 @@ export class HyperlaneSmartProvider
...
+    }
+  }
+
+  async getFeeData(): Promise<providers.FeeData> {
+    // override hardcoded getFeedata
+    const { block, gasPrice } = await resolveProperties({
+      block: this.getBlock('latest'),
+      gasPrice: this.getGasPrice().catch((error) => {
+        // @TODO: Why is this now failing on Calaveras?
+        //console.log(error);
```
Comment (jmrossy) [inline typescript/sdk/src/providers/SmartProvider/SmartProvider.ts:115]: cleanup

File: typescript/sdk/src/providers/SmartProvider/SmartProvider.ts [hunk 7]
```diff
@@ -97,6 +98,41 @@ export class HyperlaneSmartProvider
...
+        return null;
+      }),
+    });
+
+    let lastBaseFeePerGas = null,
+      maxFeePerGas = null,
+      maxPriorityFeePerGas = null;
```
Comment (jmrossy) [inline typescript/sdk/src/providers/SmartProvider/SmartProvider.ts:122]: types?

File: typescript/sdk/src/providers/SmartProvider/SmartProvider.ts [hunk 8]
```diff
@@ -97,6 +98,41 @@ export class HyperlaneSmartProvider
...
+
+    if (block && block.baseFeePerGas) {
```
Comment (jmrossy) [inline typescript/sdk/src/providers/SmartProvider/SmartProvider.ts:124]:
```suggestion
    if (block?.baseFeePerGas) {
```

File: typescript/sdk/src/providers/SmartProvider/SmartProvider.ts [hunk 9]
```diff
@@ -97,6 +98,41 @@ export class HyperlaneSmartProvider
...
+    } catch (error) {
+      return BigNumber.from('1500000000');
+    }
+  }
+
+  async getFeeData(): Promise<providers.FeeData> {
+    // override hardcoded getFeedata
+    const { block, gasPrice } = await resolveProperties({
+      block: this.getBlock('latest'),
+      gasPrice: this.getGasPrice().catch((error) => {
```
Comment (jmrossy) [inline typescript/sdk/src/providers/SmartProvider/SmartProvider.ts:113]: Are you certain we want to swallow errors here?

File: typescript/sdk/src/token/deploy.ts [hunk 10]
```diff
@@ -83,58 +83,69 @@ abstract class TokenDeployer<
...
-        ]);
-
-        return { name, symbol, totalSupply: totalSupply.toString(), decimals };
       }
-    }
 
-    return undefined;
+      return undefined;
+    } catch (error) {
+      console.error(`Could not derive token metadata`, configMap, error);
```
Comment (jmrossy) [inline typescript/sdk/src/token/deploy.ts:138]: use `this.logger`, not console

File: typescript/sdk/src/token/deploy.ts [hunk 11]
```diff
@@ -83,58 +83,69 @@ abstract class TokenDeployer<
...
-          erc20.decimals(),
-        ]);
-
-        return { name, symbol, totalSupply: totalSupply.toString(), decimals };
       }
-    }
 
-    return undefined;
+      return undefined;
+    } catch (error) {
```
Comment (jmrossy) [inline typescript/sdk/src/token/deploy.ts:137]: It may have been intentional to let this error propagate before. Are you certain the deployment should proceed even if we can't derive metadata in the way we expected to? Also these changes seem unrelated to SmartProvider and should preferably be in a separate PR

File: typescript/sdk/src/providers/SmartProvider/HyperlaneJsonRpcProvider.ts (hunk 3 above)
Comment (nambrot) [reply typescript/sdk/src/providers/SmartProvider/HyperlaneJsonRpcProvider.ts:38]: Just extending the functionality from the base json rpc provider in

https://github.com/ethers-io/ethers.js/blob/v5/packages/providers/src.ts/json-rpc-provider.ts#L555

File: typescript/sdk/src/token/deploy.ts (hunk 11 above)
Comment (yorhodes) [reply typescript/sdk/src/token/deploy.ts:137]: heavy agree with this, this is inconsistent now with how the schema assesses whether we have "sufficient" metadata to deploy

File: typescript/sdk/src/providers/SmartProvider/SmartProvider.ts (hunk 5 above)
Comment (nambrot) [reply typescript/sdk/src/providers/SmartProvider/SmartProvider.ts:105]: Doesn't this perform at the SmartProvider level? so the retries happen inside `perform`?

File: typescript/sdk/src/providers/SmartProvider/SmartProvider.ts (hunk 9 above)
Comment (nambrot) [reply typescript/sdk/src/providers/SmartProvider/SmartProvider.ts:113]: Copied from the upstream implementation and would keep it as is to avoid deviations

File: typescript/sdk/src/token/deploy.ts (hunk 10 above)
Comment (nambrot) [reply typescript/sdk/src/token/deploy.ts:138]: This is a `static` function so this.logger is not available?

File: typescript/sdk/src/token/deploy.ts (hunk 11 above)
Comment (nambrot) [reply typescript/sdk/src/token/deploy.ts:137]: Yes sorry, re-throwing the error

**Comment by nambrot** [comment]
Closing in favor of #3971

File: typescript/sdk/src/token/deploy.ts (hunk 10 above)
Comment (jmrossy) [reply typescript/sdk/src/token/deploy.ts:138]: Then use rootLogger
